.github/workflows/covector-version-or-publish.yml

```diff
@@ -48,3 +47,22 @@ jobs:
           body: ${{ steps.covector.outputs.change }}
           token: ${{ secrets.FRONTSIDEJACK_GITHUB_TOKEN }}
           author: frontsidejack frontsidejack@users.noreply.github.com
+
+  deploy-typedocs:
+    runs-on: ubuntu-latest
+    needs: version-or-publish
+    if: ${{ needs.version-or-publich.outputs.successfulPublish == 'true' }}
+    steps:
+      - uses: actions/checkout@v2
+      - run: |
+          yarn
+          yarn prepack
+          yarn docs
+      - uses: actions/cache@v2
+        id: typedoc
+        with:
+          path: docs/api/v2
+          key: typedoc-cache
```

Comment:
I'm not quite following how the cache will work. To me it looks like it is generating the docs every time on a successful publish, and only trying to restore the cache _after_ the docs are already generated. Also, because the cache key is static, it will only ever work once.

Is successful publish enough here to only have the api docs publish on a release? If I understand correctly, this will only be true when covector _actually_ publishes new versions to NPM? If that's the case, then yes, we have the possibility that the docs will publish for cases where we make a release that only has internal changes. However, that's still a fraction of releases which is a fraction of all PRS which sounds like we'd be avoiding the vast majority of the unnecessary work.

If that's not the case, then I think we'll need to:

1. restore the cache _before_ generating the docs
2. only generate the docs on a cache miss
3. make the cache key a derivative of all the public version numbers of the packages

Reply:
Yes, the point of the workflow is to prevent publishing new docs until we know the new changes have been published to npm. Otherwise we might have new changes in docs while the version-bump is still stuck in version-packages pr.

I'm not sure how we can check for cache hits without actually building the docs. Alternatively we could just omit the caching and building in github actions and let netlify take care of it.

Reply:
> I'm not sure how we can check for cache hits without actually building the docs

This is what the cache key is all about. You generate the cache key from the sources which is presumed to be inexpensive. For example, we could concatenate all the versions together, so the cache key might be:

```
@effection/core-2.0.0-beta.5;@effection/subscription-2.0.0.beta.6;@effection/main-2.0.0-beta.1;effection-2.0.0-beta.4;
```

If that changes, then you get a cache miss and need to regenerate the docs. As it stands here, the key `typedoc-cache` will only ever miss once, and will forever hit after that.

Reply:
if i'm understanding you correctly, basing it on the package versions, we won't need the cache mechanism because covector will tell us if new packages have been published

> If that changes, then you get a cache miss and need to regenerate the docs. As it stands here, the key typedoc-cache will only ever miss once, and will forever hit after that.

```
- uses: actions/cache@v2
  id: typedoc
  with:
    path: docs/api/v2          <---- 👈 
    key: typedoc-cache
```
We're caching the directory of the generated docs (`./api/docs/v2/`) so it should miss every time the docs' contents or styles have changed... ?

Reply:
> Alternatively we could just omit the caching and building in github actions and let netlify take care of it.

As long as this only ever happens after an actual publish to NPM, and not just every time we check a publish, then we should be good.

Reply:
> We're caching the directory of the generated docs (./api/docs/v2/) so it should miss every time the docs' contents or styles have changed... ?

`path` (and the files therein) is the value that is cached, `key` is what determines if there is a miss.

You can think of it as a key->value map. If the key exists in the cache, then you have a hit and pull the value. If the key does not exist then you have a miss and you need to generate (and then save) the value. In this case, the key is `typedoc-cache` which never changes, so you will always hit (except for the first time)

Reply:
ohh I see, I misused the key